Uploader/TraceFileScanner.cs

```diff
@@ -0,0 +1,147 @@
+    private readonly Regex versionAssemblyRegex;
+    private readonly IFileSystem fileSystem;
+
+    public TraceFileScanner(string traceDirectory, string versionAssembly, IFileSystem fileSystem)
+    {
+        this.traceDirectory = traceDirectory;
+        this.versionAssemblyRegex = new Regex(@"^Assembly=" + Regex.Escape(versionAssembly) + @".*Version:([^ ]*).*", RegexOptions.IgnoreCase);
+        this.fileSystem = fileSystem;
+    }
+
+    /// <summary>
+    /// Returns all trace files that can be uploaded or archived.
+    /// </summary>
+    public IEnumerable<ScannedFile> ListTraceFilesReadyForUpload()
+    {
+        List<string> files;
+        try
+        {
+            files = fileSystem.Directory.EnumerateFiles(traceDirectory).ToList();
+        }
+        catch (Exception e)
+        {
+            logger.Error(e, "Unable to list files in {traceDirectory}. Will retry later", traceDirectory);
+            yield break;
+        }
+        
+        logger.Debug("Scanning {fileCount} files", files.Count);
+        foreach (string filePath in files)
+        {
+            string fileName = Path.GetFileName(filePath);
+            if (!IsTraceFile(fileName))
+            {
+                logger.Debug("Skipping file that does not look like a trace file: {unknownFilePath}", filePath);
+                continue;
+            }
+
+            ScannedFile scannedFile = ScanFile(filePath);
+            if (scannedFile != null)
+            {
+                yield return scannedFile;
+            }
+        }
+    }
+
+    /// <summary>
+    /// Scans the given file path and returns the resulting ScannedFile or null in case the file should be ignored.
+    /// </summary>
+    private ScannedFile ScanFile(string filePath)
+    {
+        string[] lines;
+        try
+        {
+            lines = fileSystem.File.ReadAllLines(filePath);
+        }
+        catch (Exception e)
+        {
+            logger.Error(e, "Unable to read from file {trace}. Ignoring this file", filePath);
+            return null;
+        }
+
+        if (!IsFinished(lines))
+        {
+            logger.Debug("Ignoring unfinished trace {tracePath}", filePath);
+            return null;
+        }
+
+        string version = FindVersion(lines, filePath);
+        return new ScannedFile
+        {
+            FilePath = filePath,
+            Version = version,
+        };
+    }
+
+    private bool IsFinished(string[] lines)
+    {
+        return lines.Any(line => line.StartsWith("Jitted=") || line.StartsWith("Inlined="));
+    }
+
+    private string FindVersion(string[] lines, string tracePath)
+    {
+        string matchingLine = lines.FirstOrDefault(line => versionAssemblyRegex.IsMatch(line));
+        if (matchingLine == null)
+        {
+            logger.Debug("Did not find the version assembly in {tracePath}", tracePath);
+            return null;
+        }
+
+        Match match = versionAssemblyRegex.Match(matchingLine);
```

Comment:
you match twice here... lines.Select(line => versionAssemblyRegex.Match(line)).Where(match => match.IsMatch()).FirstOrDefault()